src/index.js

```diff
@@ -113,82 +107,219 @@ function renderClassComponent(vnode, context) {
+	let current = [data, 0];
+	// our stack contains the history of what we are handling and have
+	// handled while we do a depth-first traversal of the vnode-tree.
+	const stack = [current];
+	// The stringified output for the depth-first traversal of our vnode-tree
+	let output = '';
+
+	while (stack.length) {
+		// When we see that the length of our children got exceeded
+		// we know that we can go back up a level.
+		// we can use this to close dom-tags
+		if (current[1] >= current[0].length) {
+			const lastItem = current[0][current[1] - 1];
+			const nodeType = typeof lastItem.node.type;
+
+			if (nodeType === 'function' || nodeType === 'string') {
+				lastItem.node[PARENT] = undefined;
+				if (afterDiff) afterDiff(lastItem.node);
+				if (ummountHook) ummountHook(lastItem.node);
+			}
 
 			if (
-				typeof child === 'string' ||
-				typeof child === 'number' ||
-				typeof child === 'bigint'
+				!Array.isArray(lastItem.node) &&
+				typeof lastItem.parent.type === 'string'
 			) {
-				// @ts-ignore manually constructing a Text vnode
-				vnode[i] = h(null, null, child);
+				output += '</' + lastItem.parent.type + '>';
 			}
+
+			stack.pop();
+			current = stack[stack.length - 1];
+			continue;
 		}
-		return rendered;
-	}
 
-	// VNodes have {constructor:undefined} to prevent JSON injection:
-	if (vnode.constructor !== undefined) return '';
-
-	vnode[PARENT] = parent;
-	if (beforeDiff) beforeDiff(vnode);
-
-	let type = vnode.type,
-		props = vnode.props,
-		cctx = context,
-		contextType,
-		rendered,
-		component;
-
-	// Invoke rendering on Components
-	if (typeof type === 'function') {
-		if (type === Fragment) {
-			rendered = props.children;
-		} else {
-			contextType = type.contextType;
-			if (contextType != null) {
-				let provider = context[contextType.__c];
-				cctx = provider ? provider.props.value : contextType.__;
+		const item = current[0][current[1]];
+		const { node, parent } = item;
+		let { context, selectValue, isSvgMode } = item;
+
+		switch (getStep(node)) {
+			// Cases that result in null
+			case NULL_NODE_TYPE:
+			case FAUX_NODE_TYPE: {
+				current[1]++;
+				current = stack[stack.length - 1];
+				continue;
+			}
+			// We discover a new array, we have to gather all children
+			// in a shape that allows us to handle them within our iteration.
+			// this means that we will convert them to our item shape and
+			// add them to our stack as data-points.
+			case LIST_TYPE: {
+				parent[CHILDREN] = node;
+				const data = [];
+				for (let i = 0; i < node.length; i++) {
+					let child = node[i];
+					const childType = typeof child;
+					if (child == null || childType === 'boolean') continue;
+
+					data.push(
+						createStackItem(child, context, parent, isSvgMode, selectValue)
+					);
+
+					if (
+						childType === 'string' ||
+						childType === 'number' ||
+						childType === 'bigint'
+					) {
+						// @ts-ignore manually constructing a Text vnode
+						node[i] = h(null, null, child);
```

Comment:
This costs us a lot of perf in compat as we invoke `options.vnode` more frequently